### PR TITLE
Refactoring agent logging api endpoints

### PR DESCRIPTION
**Pull request by vespian**
This PR:
* introduces changes to how Agent's listen address is configured so that it's possible to run agent AR instances in the test harness
* it refactors logging API endpoints so that:
  ** fixes bug with some of the requests ending with `//` in the URL
  ** cleanups the rewrite rules
* various test-harness fixes that make testing things mentioned above possible.

Changes in DC/OS *must* follow.

### CODE REVIEW

File: nginx.master.conf [hunk 1]
```diff
@@ -285,35 +285,42 @@ http {
...
-        location ~ ^/system/v1/agent/(?<agentid>[0-9a-zA-Z-]+)/logs/v1/(?<url>.*)$ {
+        location /system/v1/metrics/ {
+            access_by_lua 'auth.validate_jwt_or_exit()';
+
+            include common/proxy-headers.conf;
+            proxy_pass http://metrics/;
+        }
+
+		# Ordering is important for these two endpoints!!!
+        location ~ ^/system/v1/agent/(?<agentid>[0-9a-zA-Z-]+)(?<type>(/metrics/v0|/logs/v1))(?<url>.*)$ {
```
Comment (malnick) [inline nginx.master.conf]: Logs requires SSE headers here and metrics does not - do we want them in the same location block?

**Comment by malnick** [comment]
Big fan of the new test infra here. I have one comment about having metrics and logs be in the same location. If was my understanding that logs has SSE and that's why it gets it's own location previously. Furthermore, I think there was a single catch all for general /agent/<agent_id> location, where this PR makes it metrics or logs specific.

File: nginx.master.conf (hunk 1 above)
Comment (vespian) [inline nginx.master.conf]: Could you please double-check ? A fragement from @philipnrmn  PR:
```
         location ~ ^/system/v1/agent/(?<agentid>[0-9a-zA-Z-]+)/logs/v1/(?<url>.*)$ {
             access_by_lua 'auth.validate_jwt_or_exit()';
             set $agentaddr '';
             more_clear_input_headers Accept-Encoding;
             rewrite ^/(slave|agent)/[0-9a-zA-Z-]+/.*$ $url break;
             rewrite_by_lua_file conf/master/agent.lua;
 
             include common/server-sent-events.conf;
             include common/proxy-headers.conf;
 
             proxy_pass http://$agentaddr:61001/system/v1/logs/v1/$url$is_args$query_string;
         }
 
(...)

        location ~ ^/system/v1/agent/(?<agentid>[0-9a-zA-Z-]+)/metrics/v0/(?<url>.*)$ {
            access_by_lua 'auth.validate_jwt_or_exit()';
            set $agentaddr '';
            more_clear_input_headers Accept-Encoding;
            rewrite ^/(slave|agent)/[0-9a-zA-Z-]+/.*$ $url break;
            rewrite_by_lua_file conf/master/agent.lua;

            include common/server-sent-events.conf;
            include common/proxy-headers.conf;

            proxy_pass http://$agentaddr:61001/system/v1/metrics/v0/$url$is_args$query_string;
        }
```

I have concatenated only `/logs/v1` and `/metrics/v0` - they seem identical. Please correct me if I am wrong.

Comment (vespian) [inline nginx.master.conf]: Maybe you mean the main agent endpoint which is still separate and does not include SSE:
```
        location ~ ^/system/v1/agent/(?<agentid>[0-9a-zA-Z-]+)(?<url>.*)$ {
            access_by_lua 'auth.validate_jwt_or_exit()';
            set $agentaddr '';
            more_clear_input_headers Accept-Encoding;
            rewrite ^/(slave|agent)/[0-9a-zA-Z-]+/.*$ $url break;
            rewrite_by_lua_file conf/master/agent.lua;

            include common/proxy-headers.conf;

            proxy_pass http://$agentaddr:61001/system/v1/$url$is_args$query_string;
}
```

Comment (malnick) [inline nginx.master.conf]: @darkonie can you double check this please? RE: your comment yesterday in the day2ops channel.

**Comment by malnick** [comment]
Ok, looks like we don't need a specific metrics location here, it can use the generalized proxy to a specific agent. Thanks @vespian

File: nginx.master.conf (hunk 1 above)
Comment (philipnrmn) [inline nginx.master.conf]: @vespian this was a copy-paste error from me - I don't believe the metrics endpoint requires SSEs.

**Comment by alberts** [comment]
:ship: but let's get @jgehrcke to take a really quick look

File: nginx.master.conf [hunk 2]
```diff
@@ -285,35 +285,42 @@ http {
...
 
-        location ~ ^/system/v1/agent/(?<agentid>[0-9a-zA-Z-]+)/logs/v1/(?<url>.*)$ {
+        location /system/v1/metrics/ {
+            access_by_lua 'auth.validate_jwt_or_exit()';
+
+            include common/proxy-headers.conf;
+            proxy_pass http://metrics/;
+        }
+
+		# Ordering is important for these two endpoints!!!
```
Comment (jgehrcke) [inline nginx.master.conf]: In the next iteration, let's dedent this comment.

Comment (vespian) [inline nginx.master.conf]: 👍

Comment (vespian) [inline nginx.master.conf]: My vim has mixed again tabs and spaces :/

File: nginx.master.conf [hunk 3]
```diff
@@ -285,35 +285,42 @@ http {
...
 
+            rewrite ^/agent/[0-9a-zA-Z-]+(.*)$ $1 break;
+
+            more_clear_input_headers Accept-Encoding;
             include common/server-sent-events.conf;
             include common/proxy-headers.conf;
 
-            proxy_pass http://$agentaddr:61001/system/v1/logs/v1/$url$is_args$query_string;
-        }
+            proxy_pass http://$agentaddr:61001/system/v1$type$url$is_args$query_string;
```
Comment (jgehrcke) [inline nginx.master.conf]: Does this `$type` variable ever have a value other than `/logs/v1`?

File: nginx.master.conf [hunk 4]
```diff
@@ -285,35 +285,42 @@ http {
...
             access_by_lua 'auth.validate_jwt_or_exit()';
+
             set $agentaddr '';
-            more_clear_input_headers Accept-Encoding;
-            rewrite ^/(slave|agent)/[0-9a-zA-Z-]+/.*$ $url break;
             rewrite_by_lua_file conf/master/agent.lua;
 
+            rewrite ^/agent/[0-9a-zA-Z-]+(.*)$ $1 break;
+
+            more_clear_input_headers Accept-Encoding;
```
Comment (jgehrcke) [inline nginx.master.conf]: Does not block merge, but I'd really love to understand why we do this header modification. Probably worth a comment in one of the next iterations.

File: nginx.master.conf (hunk 3 above)
Comment (vespian) [inline nginx.master.conf]: ATM no, but I wanted to leave a placeholder for future endpoints. Something like "Place the next prefix here".

File: nginx.master.conf (hunk 4 above)
Comment (vespian) [inline nginx.master.conf]: We can ask @darkonie

**Comment by jgehrcke** [comment]
Looks good! Would be great if we could normalize the use of indentation/tabs/spaces, but can also do this in the next iteration. :ship:

File: nginx.master.conf (hunk 3 above)
Comment (jgehrcke) [inline nginx.master.conf]: ok

**Pull request merged**